### ISSUE TITLE
[prefer-readonly-parameter-types] Support all readonly built-in types

### ISSUE THREAD

**Issue by stekycz** [typescript-eslint/typescript-eslint#1758]
I would expect the rule to accept all built-in types that are considered read-only. The rules check objects, arrays, and tuples. However, it considers `ReadonlySet` and `ReadonlyMap` as objects resulting in false positive.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/prefer-readonly-parameter-types": ["error", {
      "checkParameterProperties": true
    }]
  }
}
```

```TS
const x = (data: ReadonlySet<string>): void => {};
const y = (data: ReadonlyMap<string, string>): void => {};
```

**Expected Result**

Pass linting because the types used are read-only.

**Actual Result**

Fails because the type itself has a few methods that are not defined read-only (because they are methods, not properties with function values).

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.24.0` |
| `@typescript-eslint/parser`        | `2.24.0` |
| `TypeScript`                       | `3.8.3` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `12.15.0` |
| `yarn`                              | `1.22.4` |

**Comment by bradzacher** [comment]
The problem is that these aren't truly readonly though.
This rule makes a conscious decision to _not_ ignore methods that aren't readonly, because they can be assigned to, and thus can be a source of bugs.

[There is an exception for `ReadonlyArray<T>`/`readonly T[]`, because typescript has special handling under the hood which makes it impossible to make the methods readonly on any array-like type (i.e. `Readonly<ReadonlyArray<T>> === ReadonlyArray<T>`)](https://ts-ast-viewer.com/#code/CYUwxgNghgTiAEYD2A7AzgF3gDwFzwCUQphUIBPAQRhinIB5MYBLFAcwD4BuAKFElgJk6LOXxESZBhnIAHEEgBmObjz7hocRKkzxFSJPigpyvHtgB0wsFCwBePQd7krqG-cdJeQA)

Maybe this should be behind an option? So you can choose to opt in to some readonly "unsafety"?

**Comment by stekycz** [comment]
I understand 🤔 probably...

So the array methods might be reassigned as well but it needs extra handling because there is no other way?

I am thinking about how to make the option work. I have also created a [partially related issue for luxon's type definitions](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/43182). I see 3 possible solutions right now:

1. List of types that are considered read-only (e.g. `ReadonlySet`, `ReadonlyMap`). However, that is not usable with luxon without the issue above fixed.
2. Special flag to consider implemented list of types as read-only not allowing custom types in the list.
3. Flag to consider all object methods read-only. However, that would weaken the rule which might introduce some bugs (that would be hard to trace).

There is always a way to use `Readonly<...>` everywhere. However, that makes the code less readable IMO. For some type definitions [`DeepReadonly<...>`](https://github.com/piotrwitek/utility-types#deepreadonlyt) might be used, however, it makes the readability worse adding an extra dependency (that might be useful anyway).

I believe that using `Readonly<...>` (or [`DeepReadonly<...>`](https://github.com/piotrwitek/utility-types#deepreadonlyt) respectively) is the only correct way. From the options above, I think that the third option might be implemented because it makes sense in the context of generic usage even it might not be as safe as possible.

What is your opinion?

**Comment by bradzacher** [comment]
> So the array methods might be reassigned as well but it needs extra handling because there is no other way?

Correct, there is literally no way to create a truly readonly `ReadonlyArray` without manually extending `ReadonlyArray` in your codebase. Which is well beyond the scope of what a lint rule should enforce, so I don't think this should ever be checked.

---

I think having people bring in a pre-defined `DeepReadonly` type is fine, as there are plenty out there in the wild. `DeepReadonly` is only required if you're passing nested objects around as well.

---

I think having some options to allow a user to opt in to some "unsafety" is okay:
- `treatReadonlyBuiltinsAsSafe: boolean` could be a good way to have a pre-defined set of builtin types we can curate as required (i.e. initially would just be `ReadonlySet` and `ReadonlyMap`).
- `treatMethodsAsReadonly: boolean` could let people instruct the rule to ignore methods.

I don't really like "whitelist" style options, because they are contextless, so they can easily false-negative (causing bug vectors), so I hesitate to suggest adding an option for a list of type names to ignore.

The above two options would probably be a good start.
The first one is dead simple to implement, and could borrow code from our `no-throw-literal` rule, which checks to see if a type is the `Error` type from the typescript libraries:
https://github.com/typescript-eslint/typescript-eslint/blob/abf1a2fa5574e41af8070be3d79a886ea2f989cc/packages/eslint-plugin/src/rules/no-throw-literal.ts#L36-L44

**Comment by stekycz** [comment]
Thank you for such a response! I can definitely prepare a PR for `treatReadonlyBuiltinsAsSafe: boolean`.

-----

The second option (`treatMethodsAsReadonly: boolean`) could be treated in 2 possible way I can imagine:

1. Differentiate methods and properties (even the property type is a function type).
2. Differentiate methods and function typed properties from other properties.

I believe the first is the correct one, however, it might be harder to implement as you suggest.

-----

However, do you want to support both options? And do you want to allow them combining or should the options schema require them exclusively? I would expect that using both, either or none should be possible.

**Comment by bradzacher** [comment]
I wouldn't bother making them exclusive. There is overlap, yeah, but it's not the end of the world if someone turns them both on.

> Differentiate methods and properties (even the property type is a function type).

This would be what I'd go with, because you cannot mark methods as readonly, but you can mark properties with function types as readonly.

**Comment by RebeccaStevens** [comment]
I'm willing to work on this. Has there been any work already done?
